dali/kernels/imgproc/paste/paste_gpu.h

```diff
@@ -0,0 +1,247 @@
+    span<SampleDescriptorGPU<OutputType, InputType, ndims - 1>> out_descs,
+    span<GridCellGPU<InputType, ndims - 1>> out_grid_cells,
+    const OutListGPU<OutputType, ndims> &out,
+    const InListGPU<InputType, ndims> &in,
+    const std::vector<paste::MultiPasteSampleInput<ndims - 1>> &samples,
+    const std::vector<paste::GridCellInput<ndims - 1>> &grid, int channels) {
+  assert(out_descs.size() >= in.num_samples());
+
+  for (int i = 0; i < samples.size(); i++) {
+    auto &cpu_sample = samples[i];
+    auto &gpu_sample = out_descs[i];
+    gpu_sample.out = out[i].data;
+    gpu_sample.grid_cell_start_idx = cpu_sample.grid_cell_start_idx;
+    gpu_sample.cell_counts = cpu_sample.cell_counts;
+    gpu_sample.out_pitch[1] = out[i].shape[1];
+
+    gpu_sample.out_pitch[1] *= channels;
+  }
+
+  for (int i = 0; i < grid.size(); i++) {
+    auto &cpu_grid_cell = grid[i];
+    auto &gpu_grid_cell = out_grid_cells[i];
+
+    gpu_grid_cell.in = cpu_grid_cell.input_idx == -1 ? nullptr : in[cpu_grid_cell.input_idx].data;
+    gpu_grid_cell.cell_start = cpu_grid_cell.cell_start;
+    gpu_grid_cell.cell_end = cpu_grid_cell.cell_end;
+    gpu_grid_cell.in_anchor = cpu_grid_cell.in_anchor;
+    gpu_grid_cell.in_pitch[1] =
+            cpu_grid_cell.input_idx == -1 ? -1 : in[cpu_grid_cell.input_idx].shape[1];
+
+    gpu_grid_cell.cell_start[1] *= channels;
+    gpu_grid_cell.cell_end[1] *= channels;
+    gpu_grid_cell.in_anchor[1] *= channels;
+    gpu_grid_cell.in_pitch *= channels;
+  }
+}
+
+
+template <class OutputType, class InputType, int ndims>
+__global__ void
+PasteKernel(const SampleDescriptorGPU<OutputType, InputType, ndims> *samples,
+                  const GridCellGPU<InputType, ndims> *grid_cells,
+                  const BlockDesc<ndims> *blocks) {
+  static_assert(ndims == 2, "Function requires 2 dimensions in the input");
+  const auto &block = blocks[blockIdx.x];
+  const auto &sample = samples[block.sample_idx];
+  const GridCellGPU<InputType, ndims> *my_grid_cells = grid_cells + sample.grid_cell_start_idx;
+
+
+  auto *__restrict__ out = sample.out;
+
+  int grid_y = 0;
+
+  for (int y = threadIdx.y + block.start.y; y < block.end.y; y += blockDim.y) {
+    while (y >= my_grid_cells[grid_y * sample.cell_counts[1]].cell_end[0]) grid_y++;
+    int grid_x = 0;
+    for (int x = threadIdx.x + block.start.x; x < block.end.x; x += blockDim.x) {
+      while (x >= my_grid_cells[grid_y * sample.cell_counts[1] + grid_x].cell_end[1]) grid_x++;
+      const GridCellGPU<InputType, ndims> *cell =
+              &my_grid_cells[grid_y * sample.cell_counts[1] + grid_x];
+      if (cell->in == nullptr) {
+          out[y * sample.out_pitch[1] + x] = 0;
+      } else {
+        out[y * sample.out_pitch[1] + x] = ConvertSat<OutputType>(
+                cell->in[(y - cell->cell_start[0] + cell->in_anchor[0]) * cell->in_pitch[1]
+                         + (x - cell->cell_start[1] + cell->in_anchor[1])]);
+      }
+    }
+  }
+}
```

Comment:
This doesn't scale well - you could have some kind of block descriptor that would contain the range of contributing grid cells (possibly empty, if there are no data-containing cells). With that approach, most of the blocks would have much fewer contributing cells (typically 1).

Reply:
On the other hand, assuming that the logical block is not too large compared to inputs, you could also have just a list of contributing _inputs_ for each block - then you'd just traverse it backwards in each thread, yielding the topmost one. This is much simpler to implement and should be reasonably fast.

Reply:
Assume that the grid size is GridX by GridY cells and the current block intersects with a subgrid of size BlockGridX by BlockGridY.

In each thread the while loop for Y will run at most GridY times. 
I changed the kernel a bit. Now each thread finds minimum_grid_x before it runs, which takes GridX iterations of the loop.

Thanks to that change, though, when looping over pixels themself, `grid_x++` in the while will run `BlockGridX` times for each `y` level, instead of `GridX`, as it was in the previous solution. With this change, I think the performance should be about equal to your first solution.

Of course it has to check if it should increment `grid_x` before filling every pixel, but if I understand correctly that would also happen in your first solution, even if the contributing cell count was 1.

We can get rid of (GridX + GridY) constant by assigning min_grid_x and min_grid_y for each block, but I do not have any idea how to precompute something for each block.